src/expr/dtype_cons.cpp

```diff
@@ -0,0 +1,669 @@
+      arg = isValue ? selType.mkGroundValue() : selType.mkGroundTerm();
+    }
+    if (arg.isNull())
+    {
+      Trace("datatypes") << "...unable to construct arg of "
+                         << d_args[i]->getName() << std::endl;
+      return Node();
+    }
+    else
+    {
+      Trace("datatypes") << "...constructed arg " << arg.getType() << std::endl;
+      groundTerms.push_back(arg);
+    }
+  }
+
+  Node groundTerm = nm->mkNode(APPLY_CONSTRUCTOR, groundTerms);
+  if (isParam)
+  {
+    Assert(DType::datatypeOf(d_constructor).isParametric());
+    // type is parametric, must apply type ascription
+    Debug("datatypes-gt") << "ambiguous type for " << groundTerm
+                          << ", ascribe to " << t << std::endl;
+    groundTerms[0] = nm->mkNode(
+        APPLY_TYPE_ASCRIPTION,
+        nm->mkConst(AscriptionType(getSpecializedConstructorType(t).toType())),
+        groundTerms[0]);
+    groundTerm = nm->mkNode(APPLY_CONSTRUCTOR, groundTerms);
+  }
+  return groundTerm;
+}
+
+void DTypeConstructor::computeSharedSelectors(TypeNode domainType) const
+{
+  if (d_shared_selectors[domainType].size() < getNumArgs())
+  {
+    TypeNode ctype;
+    if (domainType.isParametricDatatype())
+    {
+      ctype = getSpecializedConstructorType(domainType);
+    }
+    else
+    {
+      ctype = d_constructor.getType();
+    }
+    Assert(ctype.isConstructor());
+    Assert(ctype.getNumChildren() - 1 == getNumArgs());
+    // compute the shared selectors
+    const DType& dt = DType::datatypeOf(d_constructor);
+    std::map<TypeNode, unsigned> counter;
+    for (unsigned j = 0; j < ctype.getNumChildren() - 1; j++)
+    {
+      TypeNode t = ctype[j];
+      Node ss = dt.getSharedSelector(domainType, t, counter[t]);
+      d_shared_selectors[domainType].push_back(ss);
+      Assert(d_shared_selector_index[domainType].find(ss)
+             == d_shared_selector_index[domainType].end());
+      d_shared_selector_index[domainType][ss] = j;
+      counter[t]++;
+    }
+  }
+}
+
+bool DTypeConstructor::resolve(
+    TypeNode self,
+    const std::map<std::string, TypeNode>& resolutions,
+    const std::vector<TypeNode>& placeholders,
+    const std::vector<TypeNode>& replacements,
+    const std::vector<TypeNode>& paramTypes,
+    const std::vector<TypeNode>& paramReplacements,
+    size_t cindex)
+{
+  if (isResolved())
+  {
+    // already resolved, fail
+    return false;
+  }
+  Trace("datatypes") << "DTypeConstructor::resolve, self type is " << self
+                     << std::endl;
+
+  NodeManager* nm = NodeManager::currentNM();
+  size_t index = 0;
+  std::vector<TypeNode> argTypes;
+  for (std::shared_ptr<DTypeSelector> arg : d_args)
+  {
+    std::string argName = arg->d_name;
+    TypeNode range;
+    if (arg->d_selector.isNull())
+    {
+      // the unresolved type wasn't created here; do name resolution
+      std::string typeName = argName.substr(argName.find('\0') + 1);
```

Comment:
If there is no other way than storing additional information in the string, at least create helper functions do extract the corresponding part of the string.